### PR TITLE
Don't require `end` on `/messages`

### PR DESCRIPTION
**Pull request by S7evinK**
As per the spec, the `end` key is not required if there are no more events for the user.

> If no further events are available (either because we have reached the start of the timeline, or because the user does not have permission to see any more events), this property is omitted from the response.

### CODE REVIEW

**Comment by DMRobertson** [comment]
I think the spec's wording is actually stricter here: I think it's saying that `end` *must* be omitted if there are no more messages to retrieve.

Should we be asserting that `end` is _not_ present in the json body?

**Comment by S7evinK** [comment]
Makes sense, yea.
Just tested that locally, Synapse isn't yet happy with that, though.

**Comment by richvdh** [comment]
I thought that recently got fixed in synapse.

**Comment by DMRobertson** [comment]
> I thought that recently got fixed in synapse.

matrix-org/synapse#12903

**Comment by S7evinK** [comment]
Hm, what happens if the returned `chunk` count is less than the requested limit? Shouldn't that also result in `end` being omitted?
Or would that be [this](https://github.com/matrix-org/synapse/pull/12903/files#diff-a3b3a9ae833ad473bff3bb3825e66d7ebdf0ef33cc1f59e4ee5fd7c030c522ecR535-R536)?

If so, we probably can't assert that `end` is actually missing in these tests, without extending the tests.

File: tests/10apidoc/34room-messages.pl [hunk 1]
```diff
@@ -158,12 +158,33 @@ sub matrix_send_room_text_message
...
 
+         # We should still get events and a "end" key, check it is actually there
          assert_json_keys( $body, qw( start end chunk ));
          assert_json_list( $body->{chunk} );
+         $token = $body->{end};
 
          scalar @{ $body->{chunk} } > 0 or
             die "Expected some messages but got none at all\n";
+      })->then( sub {
+
```
Comment (richvdh) [inline tests/10apidoc/34room-messages.pl:169]:
```suggestion
```

File: tests/10apidoc/34room-messages.pl [hunk 2]
```diff
@@ -158,12 +158,33 @@ sub matrix_send_room_text_message
...
+         # Do another call to /messages, this time we don't expect to receive a "end" key
+         do_request_json_for( $user,
+            method => "GET",
+            uri    => "/v3/rooms/$room_id/messages",
+
+            # With no params this does "forwards from END"; i.e. nothing useful
```
Comment (richvdh) [inline tests/10apidoc/34room-messages.pl:175]: I don't think this needs repeating here

```suggestion
```

File: tests/10apidoc/34room-messages.pl [hunk 3]
```diff
@@ -158,12 +158,33 @@ sub matrix_send_room_text_message
...
+         # We should still get events and a "end" key, check it is actually there
          assert_json_keys( $body, qw( start end chunk ));
          assert_json_list( $body->{chunk} );
+         $token = $body->{end};
 
          scalar @{ $body->{chunk} } > 0 or
             die "Expected some messages but got none at all\n";
+      })->then( sub {
+
+         # Do another call to /messages, this time we don't expect to receive a "end" key
```
Comment (richvdh) [inline tests/10apidoc/34room-messages.pl:170]:
```suggestion
         # Do another call to /messages. This time we don't expect to receive a "end" key
```

File: tests/10apidoc/34room-messages.pl [hunk 4]
```diff
@@ -158,12 +158,33 @@ sub matrix_send_room_text_message
       })->then( sub {
          my ( $body ) = @_;
 
+         # We should still get events and a "end" key, check it is actually there
```
Comment (richvdh) [inline tests/10apidoc/34room-messages.pl:161]:
```suggestion
         # We should still get events and a "end" key: check they are actually there
```

File: tests/10apidoc/34room-messages.pl [hunk 5]
```diff
@@ -197,10 +219,12 @@ sub matrix_send_room_text_message
          my ( $body ) = @_;
 
          log_if_fail "Body", $body;
-
+         
+         # We should still get events and a "end" key, check it is actually there
```
Comment (richvdh) [inline tests/10apidoc/34room-messages.pl:223]:
```suggestion
         # We should still get events and a "end" key: check they are actually there
```

File: tests/10apidoc/34room-messages.pl [hunk 6]
```diff
@@ -209,6 +233,25 @@ sub matrix_send_room_text_message
          scalar @{ $body->{chunk} } > 0 or
             die "Expected some messages but got none at all\n";
 
+      })->then( sub {
+         # Do another call to /messages, this time we don't expect to receive a "end" key
```
Comment (richvdh) [inline tests/10apidoc/34room-messages.pl:237]:
```suggestion
         # Do another call to /messages. This time we don't expect to receive a "end" key
```

File: tests/10apidoc/34room-messages.pl [hunk 7]
```diff
@@ -209,6 +233,25 @@ sub matrix_send_room_text_message
...
+         do_request_json_for( $user,
+            method => "GET",
+            uri    => "/v3/rooms/$room_id/messages",
+
+            # With no params this does "forwards from END"; i.e. nothing useful
```
Comment (richvdh) [inline tests/10apidoc/34room-messages.pl:242]:
```suggestion
```

**Review by richvdh — APPROVED** [review]
lgtm otherwise

**Pull request merged**